core/types/hashing.cc

```diff
@@ -0,0 +1,128 @@
+#include "core/Hashing.h"
+#include "core/SymbolRef.h"
+#include "core/Types.h"
+
+using namespace std;
+
+namespace sorbet::core {
+u4 ClassType::hash(const GlobalState &gs) const {
+    u4 value = static_cast<u4>(TypePtr::Tag::ClassType);
+    return mix(value, this->symbol.id());
+}
+
+u4 UnresolvedClassType::hash(const GlobalState &gs) const {
+    u4 result = static_cast<u4>(TypePtr::Tag::UnresolvedClassType);
+    result = mix(result, this->scope.rawId());
+    for (auto name : this->names) {
+        result = mix(result, _hash(name.shortName(gs)));
+    }
+    return result;
+}
+
+u4 UnresolvedAppliedType::hash(const GlobalState &gs) const {
+    u4 result = static_cast<u4>(TypePtr::Tag::UnresolvedAppliedType);
+    result = mix(result, this->klass.rawId());
+    for (auto &targ : targs) {
+        result = mix(result, targ.hash(gs));
+    }
+    return result;
+}
+
+u4 LiteralType::hash(const GlobalState &gs) const {
+    u4 result = static_cast<u4>(TypePtr::Tag::LiteralType);
+    auto underlying = this->underlying(gs);
+    ClassOrModuleRef undSymbol = cast_type_nonnull<ClassType>(underlying).symbol;
+    result = mix(result, undSymbol.id());
+
+    // Strings or numbers encode an extra value. True/False are completely represented by the class symbol.
+    // TODO: Casts are bad???
+    if (undSymbol == Symbols::String() || undSymbol == Symbols::Symbol()) {
+        result = mix(result, asName(gs).hash(gs));
+    } else if (undSymbol == Symbols::Integer()) {
+        u8 val = absl::bit_cast<u8>(asInteger());
+        u4 topBits = static_cast<u4>(val >> 32);
+        u4 bottomBits = static_cast<u4>(val & 0xFFFF);
+        result = mix(result, topBits);
+        result = mix(result, bottomBits);
+    } else if (undSymbol == Symbols::Float()) {
+        u8 val = absl::bit_cast<u8>(asFloat());
+        u4 topBits = static_cast<u4>(val >> 32);
+        u4 bottomBits = static_cast<u4>(val & 0xFFFF);
```

Comment:
0xFFFF_FFFF here as well?  Or maybe early return in the string/symbol case and merge the integer/float cases (at least once you've extracted the u8 from them)?

Reply:
Good idea. Fixed.